### ISSUE TITLE
How to run inference on an image from the wild?

### ISSUE THREAD

**Issue by LastTargaryen** [MengyuWang826/SegRefiner#15]
I'm sorry if it's like, RIGHT in front of my face and I just can't see it, but can someone please tell me how to just run plain inference on an image? I have been reviewing this entire repo for the last day and a half and I cannot figure it out. Please tell me exactly how to just run inference on an image--an image that isn't part of a dataset.

Edit: After further review, it seems I'm asking the wrong question; updated question is:

How does one extract only the mask from the result of a SAM model in order to feed it into SegRefiner...?

**Comment by xuyu666** [comment]
It's the same problem.

**Comment by LastTargaryen** [comment]
Okay, then can I please please be blessed with simple instructions for inference? I understand now I need an original image (reference image) and a single-channel mask. I have these--how do I apply them to your code?

**Comment by Alexankharin** [comment]
Inference code:
```
import numpy as np
from mmcv import Config, DictAction
from mmdet.models import build_detector


cfg=Config.fromfile('../SegRefiner/configs/segrefiner/segrefiner_hr.py')
cfg.model.type="SegRefinerSemantic"
model = build_detector(
        cfg.model,
        train_cfg=cfg.get('train_cfg'),
        test_cfg=cfg.get('test_cfg'))
model.init_weights()
model.load_state_dict(torch.load('../SegRefiner/notebooks/segrefiner_hr_latest.pth'))
model=model.eval()

def predict(img:np.ndarray, mask:np.ndarray, model:torch.nn.Module, device='cuda'):
    """
    mask 0-1 HW
    img 0-1 HWC
    """
    img_metas=[{}]
    img_metas[0]['ori_filename']='im.jpg'
    img_metas[0]['ori_shape']=[256,256,3]
    patchobj=[torch.zeros(1)]
    output=np.zeros((1024,1024))
    for h in range(4):
        for w in range(4):
            patch=img[256*h:256*(h+1),256*w:256*(w+1),:]
            maskpatch=mask[256*h:256*(h+1),256*w:256*(w+1)]
            with torch.no_grad():
                patchtorch=torch.tensor(patch.transpose(2,0,1),dtype=torch.float32).unsqueeze(0).to(device)/255
                patchobj[0].masks=(maskpatch>0)[None,...]*1
                output[256*h:256*(h+1),256*w:256*(w+1)]=model.simple_test_semantic(img_metas=img_metas, img=patchtorch, coarse_masks=patchobj, return_loss=False)[0][0]            
    return output
    
# here is we got image in HWC format and boolean mask in HW format
out=predict(np.array(image)/255,maskinit*1.0,model.cuda())
```
out is numpy array. Here i used patched-based refinement due to memory limitations for 1024x1024 image, feel free to change it

**Comment by chandana-21** [comment]
this doesn't work for me, are you sure this is '../SegRefiner/configs/segrefiner/segrefiner_hr.py'? because i didn't find these configs in there

**Comment by chandana-21** [comment]
@LastTargaryen did you find a solution yet?

**Comment by Alexankharin** [comment]
path to config should point to this file https://github.com/MengyuWang826/SegRefiner/blob/main/configs/segrefiner/segrefiner_hr.py 
path to model to the segrefiner_hr_latest.pth downloaded from https://drive.google.com/file/d/143kerk4WOerGZMqR-cAETb7rOcoQ0CQ_/view?usp=sharing

**Comment by xm1085** [comment]
> 推理代码：
> 
> ```
> import numpy as np
> from mmcv import Config, DictAction
> from mmdet.models import build_detector
> 
> 
> cfg=Config.fromfile('../SegRefiner/configs/segrefiner/segrefiner_hr.py')
> cfg.model.type="SegRefinerSemantic"
> model = build_detector(
>         cfg.model,
>         train_cfg=cfg.get('train_cfg'),
>         test_cfg=cfg.get('test_cfg'))
> model.init_weights()
> model.load_state_dict(torch.load('../SegRefiner/notebooks/segrefiner_hr_latest.pth'))
> model=model.eval()
> 
> def predict(img:np.ndarray, mask:np.ndarray, model:torch.nn.Module, device='cuda'):
>     """
>     mask 0-1 HW
>     img 0-1 HWC
>     """
>     img_metas=[{}]
>     img_metas[0]['ori_filename']='im.jpg'
>     img_metas[0]['ori_shape']=[256,256,3]
>     patchobj=[torch.zeros(1)]
>     output=np.zeros((1024,1024))
>     for h in range(4):
>         for w in range(4):
>             patch=img[256*h:256*(h+1),256*w:256*(w+1),:]
>             maskpatch=mask[256*h:256*(h+1),256*w:256*(w+1)]
>             with torch.no_grad():
>                 patchtorch=torch.tensor(patch.transpose(2,0,1),dtype=torch.float32).unsqueeze(0).to(device)/255
>                 patchobj[0].masks=(maskpatch>0)[None,...]*1
>                 output[256*h:256*(h+1),256*w:256*(w+1)]=model.simple_test_semantic(img_metas=img_metas, img=patchtorch, coarse_masks=patchobj, return_loss=False)[0][0]            
>     return output
>     
> # here is we got image in HWC format and boolean mask in HW format
> out=predict(np.array(image)/255,maskinit*1.0,model.cuda())
> ```
> 
> out 是 numpy 内存。这里我使用了基于补丁的细化，因为 1024x1024 图像的内存限制，请随意更改它

你好，为什么我对图像进行分块之后显示错误

**Comment by xm1085** [comment]
> 配置路径指向此文件[https://github.com/MengyuWang826/SegRefiner/blob/main/configs/segrefiner/segrefiner_hr.py模型路径指向从](https://github.com/MengyuWang826/SegRefiner/blob/main/configs/segrefiner/segrefiner_hr.py)[https://drive.google.com/file/d/143kerk4WOerGZMqR-cAETb7rOcoQ0CQ_ /view?usp=sharing](https://drive.google.com/file/d/143kerk4WOerGZMqR-cAETb7rOcoQ0CQ_/view?usp=sharing) 下载的 segrefiner_hr_latest.pth[](https://drive.google.com/file/d/143kerk4WOerGZMqR-cAETb7rOcoQ0CQ_/view?usp=sharing)

modelsize和img_metas和cfg.object_size和原始图像的尺寸之间的关系是什么

**Comment by Muran20** [comment]
> 推理代码：
> 
> ```
> import numpy as np
> from mmcv import Config, DictAction
> from mmdet.models import build_detector
> 
> 
> cfg=Config.fromfile('../SegRefiner/configs/segrefiner/segrefiner_hr.py')
> cfg.model.type="SegRefinerSemantic"
> model = build_detector(
>         cfg.model,
>         train_cfg=cfg.get('train_cfg'),
>         test_cfg=cfg.get('test_cfg'))
> model.init_weights()
> model.load_state_dict(torch.load('../SegRefiner/notebooks/segrefiner_hr_latest.pth'))
> model=model.eval()
> 
> def predict(img:np.ndarray, mask:np.ndarray, model:torch.nn.Module, device='cuda'):
>     """
>     mask 0-1 HW
>     img 0-1 HWC
>     """
>     img_metas=[{}]
>     img_metas[0]['ori_filename']='im.jpg'
>     img_metas[0]['ori_shape']=[256,256,3]
>     patchobj=[torch.zeros(1)]
>     output=np.zeros((1024,1024))
>     for h in range(4):
>         for w in range(4):
>             patch=img[256*h:256*(h+1),256*w:256*(w+1),:]
>             maskpatch=mask[256*h:256*(h+1),256*w:256*(w+1)]
>             with torch.no_grad():
>                 patchtorch=torch.tensor(patch.transpose(2,0,1),dtype=torch.float32).unsqueeze(0).to(device)/255
>                 patchobj[0].masks=(maskpatch>0)[None,...]*1
>                 output[256*h:256*(h+1),256*w:256*(w+1)]=model.simple_test_semantic(img_metas=img_metas, img=patchtorch, coarse_masks=patchobj, return_loss=False)[0][0]            
>     return output
>     
> # here is we got image in HWC format and boolean mask in HW format
> out=predict(np.array(image)/255,maskinit*1.0,model.cuda())
> ```
> 
> out 是 numpy 数组。这里我使用了基于补丁的细化，因为 1024x1024 图像的内存限制，请随意更改它

可以提供配置文件吗
